### ISSUE TITLE
Colorbar size not scaled with figsize

### ISSUE THREAD

**Issue by emdupre** [nilearn/nilearn#3183]
[This recent NeuroStars post ](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain) suggests that we are not properly supporting colorbars with the newest matplotlib release (3.5.1). Please see the post for an example rendering.

Nilearn version: 0.9.0

### Expected behavior
The colorbar size would be appropriately linked to the figsize.

### Actual behavior
The colorbar size does not scale with the figure.

### Steps and code to reproduce bug

```python
plotting.plot_stat_map(img, cut_coords=[20, -50, -8], colorbar = True, draw_cross = False, cmap = ‘cyan_copper’, vmax = 7)
```

**Comment by NicolasGensollen** [comment]
Indeed, I can reproduce the bug. Funny that nobody noticed it before since 3.5.1 is already 3 months old.
Also, I think the link to the neurostars post is not correct. Here is the [right link](https://neurostars.org/t/a-strange-bug-with-a-huge-colorbar/21903).

**Comment by NicolasGensollen** [comment]
@RaphaelMeudec I haven't investigated, but indeed, it seems that your code isn't producing the error...
I still get the error when running this (Nilearn main and Matplotlib 3.5.1):

```python
In [1]: from nilearn.datasets import load_mni152_template

In [2]: from nilearn import plotting

In [3]: img = load_mni152_template()
FutureWarning: Default resolution of the MNI template will change from 2mm to 1mm in version 0.10.0

In [4]: plotting.plot_stat_map(img, cut_coords=[20, -50, -8], colorbar = True, draw_cross = False, vmax = 7)
Out[4]: <nilearn.plotting.displays._slicers.OrthoSlicer at 0x7fab9804a880>

In [5]: plotting.show()
```

**Comment by jeromedockes** [comment]
the problem happens here:

  https://github.com/nilearn/nilearn/blob/183cac7b29e58f25ed8f800892209e713de4d0a7/nilearn/plotting/img_plotting.py#L256

when the colorbar is cropped (showing only the half of the colormap that is used when the image has only ≤ 0 or only ≥ 0 values), the axes containing the colorbar is modified and then so are the spines (the black lines around the axes).

But since
  https://github.com/matplotlib/matplotlib/pull/20054

the spines now use the axes transform (coordinates scaled to be between 0 and 1 rather than data coordinates):

  https://github.com/matplotlib/matplotlib/blob/8c94a2204739a9ce0d32345cacf0f39f3e0bc7b2/lib/matplotlib/colorbar.py#L200

so now if the max image value is 10 we end up with an outline that is 10 times the height of the figure.

We *can* change the outline xy coordinates we pass and use the appropriate transform for newer matplotlib versions.

However, this nilearn crop_colorbar function would then have three branches depending on matplotlib versions (and more are likely to creep in in the future). Moreover it already uses undocumented as well as explicitly private attributes of matplotlib colorbars. This suggests we should probably change the way we crop the colorbar altogether.

Rather than modifying its parts in place, maybe we can remove the existing colorbar and create a new one with the appropriate colormap, or even use the correct colormap (ranging to 0 to vmax when the image has only non-negative values) from the start (ie when plotting the brain image) and then we would not need to crop the colorbar at all.

**Comment by emdupre** [comment]
Just to be explicit, then, it seems that @RaphaelMeudec is not able to reproduce the bug (but @NicolasGensollen was) because of the `vmax` argument. In terms of a solution:

> Rather than modifying its parts in place, maybe we can remove the existing colorbar and create a new one with the appropriate colormap, or even use the correct colormap (ranging to 0 to vmax when the image has only non-negative values) from the start (ie when plotting the brain image) and then we would not need to crop the colorbar at all.

Thanks for such a nice summary, @jeromedockes ! I (personally) think this path forward sounds reasonable.

**Comment by RaphaelMeudec** [comment]
@emdupre I am not sure if the `vmax` argument is sufficient in itself. I think an important element is that the fetch_motor task gives a symmetric colorbar which does not trigger the `.set_xy` call inside the `_crop_colorbar` method as explained @jeromedockes. Plotting the mask lead to a positive-only colorbar where `_crop_colorbar` has been called.

I agree with simplifying and getting read of the matplotlib version checks, I can look into it!
